src/sys/socket/sockopt.rs

```diff
@@ -1001,6 +1001,133 @@ sockopt_impl!(
+    SetOnly,
+    libc::SOL_NETLINK,
+    libc::NETLINK_ADD_MEMBERSHIP,
+    libc::c_uint
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Drop socket from group
+    NetlinkDropMembership,
+    SetOnly,
+    libc::SOL_NETLINK,
+    libc::NETLINK_DROP_MEMBERSHIP,
+    libc::c_uint
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Enable packet info control messages
+    NetlinkPacketInfo,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_PKTINFO,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Send broadcast errors
+    NetlinkBroadcastError,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_BROADCAST_ERROR,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Avoid receiving 'ENOBUFS' errors
+    NetlinkNoEnobufs,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_NO_ENOBUFS,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Listen to notifications from all network namespaces
+    NetlinkListenAllNsid,
+    SetOnly,
+    libc::SOL_NETLINK,
+    libc::NETLINK_LISTEN_ALL_NSID,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Trim the extack if needed
+    NetlinkCapAck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_CAP_ACK,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Always send extack
+    NetlinkExtAck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_EXT_ACK,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Get strict check
+    NetlinkGetStrictCheck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_GET_STRICT_CHK,
+    bool
+);
+
+/// List netlink groups
+/// The returned Vec contains a bit set in chunks of u32.
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[derive(Copy, Clone, Debug)]
+pub struct NetlinkListMemberships;
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+impl GetSockOpt for NetlinkListMemberships {
+    type Val = Vec<u32>;
+
+    fn get(&self, fd: RawFd) -> Result<Self::Val> {
+        let mut len = 0;
+        unsafe {
+            // Get membership length
+            let res = libc::getsockopt(
+                fd,
+                libc::SOL_NETLINK,
+                libc::NETLINK_LIST_MEMBERSHIPS,
+                ::std::ptr::null_mut(),
+                &mut len,
+            );
+            Errno::result(res)?;
+            let mut val = vec![0u32; (len as usize) / mem::size_of::<u32>()];
+            if val.capacity() > 0 {
+                let old_len = len;
+                let res = libc::getsockopt(
+                    fd,
+                    libc::SOL_NETLINK,
+                    libc::NETLINK_LIST_MEMBERSHIPS,
+                    val.as_mut_ptr() as *mut libc::c_void,
+                    &mut len,
```

Comment:
This is a buffer overrun,  assuming that the `optlen` argument is the size of the array in elements, as the man page suggests.  You're telling `getsockopt` here that the array is actually four times larger than what you created on line 1109.  But you seem to be interpreting `optval` as the size of the array in bytes.  Could you find a better reference to clear up the ambiguity?

Reply:
`len` is not used as a size indicator in the kernel, it's only a out pointer.
That's why we allocate for the length first on line 1109.
The `len` returns in bytes. As it's a bitfield it's up to the user to how interpret it (u8, u16, u32...).

https://elixir.bootlin.com/linux/latest/source/net/netlink/af_netlink.c#L1754

There was a bug in this specific routine where it was under reporting the size of the array.
The maintainers argued that reporting in bytes is a better API. This of course affected code that assumed it was returning u32, but it was harmless.

https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=f4e4534850a9d18c250a93f8d7fbb51310828110

https://github.com/systemd/systemd/blob/main/src/libsystemd/sd-netlink/netlink-socket.c#L37